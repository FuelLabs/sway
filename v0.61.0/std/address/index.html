<!DOCTYPE html><html><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="swaydoc"><meta name="description" content="API documentation for the Sway `address` module in `std`."><meta name="keywords" content="sway, swaylang, sway-lang, address"><link rel="icon" href="../../static.files/sway-logo.svg"><title>address in std - Sway</title><link rel="stylesheet" type="text/css" href="../../static.files/normalize.css"><link rel="stylesheet" type="text/css" href="../../static.files/swaydoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../static.files/ayu.css"><link rel="stylesheet" href="../../static.files/ayu.min.css"></head><body class="swaydoc mod"><nav class="sidebar"><a class="sidebar-logo" href="../../std/index.html"><div class="logo-container"><img class="sway-logo" src="../../static.files/sway-logo.svg" alt="logo"></div></a><h2 class="location">Module address</h2><div class="sidebar-elems"><section><h3><a href="#structs">Structs</a></h3><ul class="block method"><li><a href="struct.Address.html">Address</a></li></ul></section></div></nav><main><div class="width-limiter"><script src="../../search.js" type="text/javascript"></script><script>function onSearchFormSubmit(event){event.preventDefault();const searchQuery=document.getElementById("search-input").value;const url=new URL(window.location.href);if(searchQuery){url.searchParams.set('search',searchQuery)}else{url.searchParams.delete('search')}history.pushState({search:searchQuery},"",url);window.dispatchEvent(new HashChangeEvent("hashchange"))}document.addEventListener('DOMContentLoaded',()=>{const searchbar=document.getElementById("search-input");const searchForm=document.getElementById("search-form");searchbar.addEventListener("keyup",function(event){searchForm.dispatchEvent(new Event('submit'))});searchbar.addEventListener("search",function(event){searchForm.dispatchEvent(new Event('submit'))});function onQueryParamsChange(){const searchParams=new URLSearchParams(window.location.search);const query=searchParams.get("search");const searchSection=document.getElementById('search');const mainSection=document.getElementById('main-content');const searchInput=document.getElementById('search-input');if(query){searchInput.value=query;const results=Object.values(SEARCH_INDEX).flat().filter(item=>{const lowerQuery=query.toLowerCase();return item.name.toLowerCase().includes(lowerQuery)});const header=`<h1>Results for ${query}</h1>`;if(results.length>0){const resultList=results.map(item=>{const formattedName=`<span class="type ${item.type_name}">${item.name}</span>`;const name=[...item.module_info,formattedName].join("::");const path=["../..",...item.module_info,item.html_filename].join("/");const left=`<td><span>${name}</span></td>`;const right=`<td><p>${item.preview}</p></td>`;return`<tr onclick="window.location='${path}';">${left}${right}</tr>`}).join('');searchSection.innerHTML=`${header}<table>${resultList}</table>`}else{searchSection.innerHTML=`${header}<p>No results found.</p>`}searchSection.setAttribute("class","search-results");mainSection.setAttribute("class","content hidden")}else{searchSection.setAttribute("class","search-results hidden");mainSection.setAttribute("class","content")}}window.addEventListener('hashchange',onQueryParamsChange);onQueryParamsChange()})</script><nav class="sub"><form id="search-form" class="search-form" onsubmit="onSearchFormSubmit(event)"><div class="search-container"><input id="search-input" class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Search the docs..." type="search"></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn"><span class="in-band">Module <a class="mod" href="../index.html">std</a><span>::</span><a class="mod" href="#">address</a></span></h1></div><details class="swaydoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>The <code>Address</code> type used for interacting with addresses on the fuel network.</p>
</div></details><h2 id="structs">Structs</h2><div class="item-table"><div class="item-row"><div class="item-left Struct-item"><a class="struct" href="struct.Address.html">Address</a></div><div class="item-right docblock-short"><p>The <code>Address</code> type, a struct wrapper around the inner <code>b256</code> value.</p>
</div></div></div></section><section id="search" class="search-results"></section></div></main><script src="../../static.files/highlight.js"></script><script>hljs.highlightAll();</script></body></html>