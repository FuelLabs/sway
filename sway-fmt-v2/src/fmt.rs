use crate::utils::{
    indent_style::Shape, newline_style::apply_newline_style, program_type::insert_program_type,
};
use std::{path::Path, sync::Arc};
use sway_core::BuildConfig;

pub use crate::{
    config::manifest::Config,
    error::{ConfigError, FormatterError},
};

#[derive(Debug, Default)]
pub struct Formatter {
    pub shape: Shape,
    pub config: Config,
}

pub type FormattedCode = String;

pub trait Format {
    fn format(
        &self,
        formatted_code: &mut FormattedCode,
        formatter: &mut Formatter,
    ) -> Result<(), FormatterError>;
}

impl Formatter {
    pub fn from_dir(dir: &Path) -> Result<Self, ConfigError> {
        let config = match Config::from_dir(dir) {
            Ok(config) => config,
            Err(ConfigError::NotFound) => Config::default(),
            Err(e) => return Err(e),
        };
        let shape = Shape::default();
        Ok(Self { config, shape })
    }
    pub fn format(
        &mut self,
        src: Arc<str>,
        build_config: Option<&BuildConfig>,
    ) -> Result<FormattedCode, FormatterError> {
        let path = build_config.map(|build_config| build_config.canonical_root_module());
        let src_len = src.len();
        let module = sway_parse::parse_file(src, path)?;
        // Get parsed items
        let items = module.items;
        // Get the program type (script, predicate, contract or library)
        let program_type = module.kind;

        // Formatted code will be pushed here with raw newline stlye.
        // Which means newlines are not converted into system-specific versions until `apply_newline_style()`.
        // Use the length of src as a hint of the memory size needed for `raw_formatted_code`,
        // which will reduce the number of reallocations
        let mut raw_formatted_code = String::with_capacity(src_len);

        // Insert program type to the formatted code.
        insert_program_type(&mut raw_formatted_code, program_type)?;

        // Insert parsed & formatted items into the formatted code.
        let mut iter = items.iter().peekable();
        while let Some(item) = iter.next() {
            // format Annotated<ItemKind>
            item.format(&mut raw_formatted_code, self)?;
            if iter.peek().is_some() {
                raw_formatted_code.push('\n');
            }
        }

        let mut formatted_code = String::from(&raw_formatted_code);
        // Replace newlines with specified `NewlineStyle`
        apply_newline_style(
            self.config.whitespace.newline_style,
            &mut formatted_code,
            &raw_formatted_code,
        )?;
        if formatted_code.ends_with('\n') {
            formatted_code.pop();
        }

        Ok(formatted_code)
    }
}

#[cfg(test)]
mod tests {
    use super::Formatter;
    use crate::config::user_def::FieldAlignment;
    use std::sync::Arc;

    #[test]
    fn test_const() {
        let sway_code_to_format = r#"contract;
pub const TEST:u16=10;"#;
        let correct_sway_code = r#"contract;

pub const TEST: u16 = 10;"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }

    #[test]
    fn test_struct_multiline_line_alignment() {
        let sway_code_to_format = r#"contract;
pub struct Foo<T, P> {
   barbazfoo: u64,
   baz  : bool,
}
"#;
        let correct_sway_code = r#"contract;

pub struct Foo<T, P> {
    barbazfoo : u64,
    baz       : bool,
}"#;

        let mut formatter = Formatter::default();
        formatter.config.structures.field_alignment = FieldAlignment::AlignFields(40);
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_struct_single_line() {
        let sway_code_to_format = r#"contract;
pub struct Foo {
    bar: u64,
    baz: bool,
}
"#;
        let correct_sway_code = r#"contract;

pub struct Foo { bar: u64, baz: bool }"#;
        let mut formatter = Formatter::default();
        formatter.config.structures.small_structures_single_line = true;
        formatter.config.whitespace.max_width = 300;
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_enum_single_line() {
        let sway_code_to_format = r#"contract;
pub enum Foo {
    bar: u64,
    baz: bool,
}
"#;
        let correct_sway_code = r#"contract;

pub enum Foo { bar: u64, baz: bool }"#;
        let mut formatter = Formatter::default();
        formatter.config.structures.small_structures_single_line = true;
        formatter.config.whitespace.max_width = 300;
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_struct_multi_line() {
        let sway_code_to_format = r#"contract;
pub struct Foo {
    bar: u64,
    baz: bool
}
"#;
        let correct_sway_code = r#"contract;

pub struct Foo {
    bar: u64,
    baz: bool,
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }

    #[test]
    fn test_enum_without_variant_alignment() {
        let sway_code_to_format = r#"contract;

enum Color {
    Blue: (), Green: (),
            Red: (),
    Silver: (),
                    Grey: () }
        "#;
        let correct_sway_code = r#"contract;

enum Color {
    Blue: (),
    Green: (),
    Red: (),
    Silver: (),
    Grey: (),
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_enum_with_variant_alignment() {
        let sway_code_to_format = r#"contract;

enum Color {
    Blue: (), Green: (),
            Red: (),
    Silver: (),
                    Grey: (), }
        "#;
        let correct_sway_code = r#"contract;

enum Color {
    Blue   : (),
    Green  : (),
    Red    : (),
    Silver : (),
    Grey   : (),
}"#;

        // Creating a config with enum_variant_align_threshold that exceeds longest variant length
        let mut formatter = Formatter::default();
        formatter.config.structures.field_alignment = FieldAlignment::AlignFields(20);

        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_item_abi_with_generics_and_attributes() {
        let sway_code_to_format = r#"contract;

abi StorageMapExample {
    #[storage(write)]fn insert_into_map1(key: u64, value: u64);

fn hello(key: u64, value: u64);
}"#;
        let correct_sway_code = r#"contract;

abi StorageMapExample {
    #[storage(write)]
    fn insert_into_map1(key: u64, value: u64);

    fn hello(key: u64, value: u64);
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }

    #[test]
    fn test_multi_items() {
        let sway_code_to_format = r#"contract;

pub const TEST: u16 = 10;
pub const TEST1: u16 = 10;"#;
        let correct_sway_code = r#"contract;

pub const TEST: u16 = 10;
pub const TEST1: u16 = 10;"#;

        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_ty_formatting() {
        let sway_code_to_format = r#"contract;

enum TestTy {
    Infer:
    _,
    Array : [u8;
    40],
    String:         str[
    4
    ],
    PathType     : root::
example::
    type,
    TupleNil: (),
    Tuple: (   u64, 
        u32
    ),
}"#;
        let correct_sway_code = r#"contract;

enum TestTy {
    Infer: _,
    Array: [u8; 40],
    String: str[4],
    PathType: root::example::type,
    TupleNil: (),
    Tuple: (u64, u32),
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code);
    }
    fn test_storage_without_alignment() {
        let sway_code_to_format = r#"contract;

storage{foo:Test=Test{},bar
: 
    Test=Test{}
, baz: u64 } 
"#;
        let correct_sway_code = r#"contract;

storage {
    foo: Test,
    bar: Test,
    baz: u64,
}"#;

        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_storage_with_alignment() {
        let sway_code_to_format = r#"contract;

storage {
 long_var_name: Type1=Type1{},
      var2: Type2=Type2{},
}
"#;
        let correct_sway_code = r#"contract;

storage {
    long_var_name : Type1,
    var2          : Type2,
}"#;

        let mut formatter = Formatter::default();
        formatter.config.structures.field_alignment = FieldAlignment::AlignFields(50);
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_storage_single_line() {
        let sway_code_to_format = r#"contract;

storage {
 long_var_name: Type1=Type1{},
      var2: Type2=Type2{},
}
"#;
        let correct_sway_code = r#"contract;

storage { long_var_name: Type1, var2: Type2 }"#;
        let mut formatter = Formatter::default();
        formatter.config.structures.small_structures_single_line = true;
        formatter.config.whitespace.max_width = 300;
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_item_fn() {
        let sway_code_to_format = r#"contract;

pub fn hello( person: String ) -> String {let greeting = 42;greeting.to_string()}
fn goodbye() -> usize {let farewell: usize = 5; farewell }"#;
        let correct_sway_code = r#"contract;

pub fn hello(person: String) -> String {
    let greeting = 42;
    greeting.to_string()
}

fn goodbye() -> usize {
    let farewell: usize = 5;
    farewell
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_same_line_where() {
        let sway_code_to_format = r#"contract;

pub fn hello( person: String ) -> String where T: Eq,{let greeting = 42;greeting.to_string()}"#;
        let correct_sway_code = r#"contract;

pub fn hello(person: String) -> String
where
    T: Eq,
{
    let greeting = 42;
    greeting.to_string()
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_trait_and_super_trait() {
        let sway_code_to_format = r#"library traits;

trait Person{ fn name( self )->String;fn age( self )->usize; }
trait Student:Person {fn university(self) -> String;}
trait Programmer {fn fav_language(self) -> String;}
trait CompSciStudent: Programmer+Student {fn git_username(self) -> String;}"#;
        let correct_sway_code = r#"library traits;

trait Person {
    fn name(self) -> String;

    fn age(self) -> usize;
}

trait Student: Person {
    fn university(self) -> String;
}

trait Programmer {
    fn fav_language(self) -> String;
}

trait CompSciStudent: Programmer + Student {
    fn git_username(self) -> String;
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
}
