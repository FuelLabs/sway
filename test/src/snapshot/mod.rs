use anyhow::Result;
use libtest_mimic::{Arguments, Trial};
use normalize_path::NormalizePath;
use regex::Regex;
use std::{
    path::{Path, PathBuf},
    str::FromStr,
    sync::Once,
};

static FORC_COMPILATION: Once = Once::new();
static FORC_DOC_COMPILATION: Once = Once::new();

fn compile_forc() {
    let args = vec!["b", "--release", "-p", "forc"];
    let o = std::process::Command::new("cargo")
        .args(args)
        .output()
        .unwrap();
    assert!(o.status.success());
}

fn compile_forc_doc() {
    let args = vec!["b", "--release", "-p", "forc-doc"];
    let o = std::process::Command::new("cargo")
        .args(args)
        .output()
        .unwrap();
    assert!(o.status.success());
}

pub(super) async fn run(filter_regex: Option<&regex::Regex>) -> Result<()> {
    let repo_root = PathBuf::from_str(&std::env::var("CARGO_MANIFEST_DIR")?)?
        .parent()
        .unwrap()
        .to_path_buf();
    let test_root = repo_root.join("test");
    let test_programs_dir = test_root.join("src/e2e_vm_tests/test_programs/");

    let args = Arguments {
        filter: filter_regex.as_ref().map(|filter| filter.to_string()),
        nocapture: true,
        ..Default::default()
    };

    let tests = discover_tests(&test_root)
        .into_iter()
        .map(|dir| {
            let name = dir
                .strip_prefix(&test_programs_dir)
                .unwrap()
                .display()
                .to_string();

            let repo_root = repo_root.clone();
            Trial::test(name, move || {
                let snapshot_toml =
                    std::fs::read_to_string(format!("{}/snapshot.toml", dir.display()))?;
                let snapshot_toml = toml::from_str::<toml::Value>(&snapshot_toml)?;
                let cmds = if let Some(cmds) = snapshot_toml.get("cmds") {
                    cmds.as_array()
                        .unwrap()
                        .iter()
                        .map(|v| v.as_str().unwrap())
                        .collect::<Vec<_>>()
                } else {
                    vec!["forc build --path {root}"]
                };

                let root = dir.strip_prefix(&repo_root).unwrap().display().to_string();

                use std::fmt::Write;
                let mut snapshot = String::new();

                for cmd in cmds {
                    let cmd = cmd.replace("{root}", &root);

                    let _ = writeln!(&mut snapshot, "> {}", cmd);

                    let mut last_output: Option<String> = None;

                    for cmd in cmd.split("|") {
                        let cmd = cmd.trim();

                        // known commands
                        let cmd = if let Some(cmd) = cmd.strip_prefix("forc doc ") {
                            FORC_DOC_COMPILATION.call_once(|| {
                                compile_forc_doc();
                            });
                            format!("target/release/forc-doc {cmd} 1>&2")
                        } else if let Some(cmd) = cmd.strip_prefix("forc ") {
                            FORC_COMPILATION.call_once(|| {
                                compile_forc();
                            });
                            format!("target/release/forc {cmd} 1>&2")
                        } else if let Some(cmd) = cmd.strip_prefix("grep ") {
                            let arg = cmd.trim();
                            if let Some(l) = last_output.take() {
                                let mut new_output = String::new();
                                for line in l.lines() {
                                    if line.contains(arg) {
                                        new_output.push_str(line);
                                        new_output.push('\n');
                                    }
                                }
                                last_output = Some(new_output);
                            }
                            continue;
                        } else {
                            panic!("Not supported. Possible commands: forc")
                        };

                        let o = duct::cmd!("bash", "-c", cmd.clone())
                            .dir(repo_root.clone())
                            .stderr_to_stdout()
                            .stdout_capture();

                        let o = if let Some(last_output) = last_output.as_ref() {
                            o.stdin_bytes(last_output.as_bytes())
                        } else {
                            o
                        };

                        let o = o.env("COLUMNS", "10").unchecked().start().unwrap();
                        let o = o.wait().unwrap();
                        last_output = Some(clean_output(&format!(
                            "exit status: {}\noutput:\n{}",
                            o.status.code().unwrap(),
                            std::str::from_utf8(&o.stdout).unwrap(),
                        )));
                    }

                    let _ = writeln!(&mut snapshot, "{}", last_output.unwrap_or_default());
                }

                fn stdout(root: &str, snapshot: &str) {
                    let root = PathBuf::from_str(root).unwrap();
                    let root = root.normalize();

                    let mut insta = insta::Settings::new();
                    insta.set_snapshot_path(root);
                    insta.set_prepend_module_to_snapshot(false);
                    insta.set_omit_expression(true);
                    let scope = insta.bind_to_scope();
                    insta::assert_snapshot!("stdout", snapshot);
                    drop(scope);
                }
                stdout(&format!("{}/{root}", repo_root.display()), &snapshot);

                Ok(())
            })
        })
        .collect();

    libtest_mimic::run(&args, tests).exit();
}

pub fn discover_tests(test_root: &Path) -> Vec<PathBuf> {
    use glob::glob;

    let mut entries = vec![];

    let pattern = format!("{}/**/snapshot.toml", test_root.display());
    for entry in glob(&pattern)
        .expect("Failed to read glob pattern")
        .flatten()
    {
        entries.push(entry.parent().unwrap().to_owned())
    }

    entries
}

fn clean_output(output: &str) -> String {
    #[derive(Default)]
    struct RawText(String);

    impl vte::Perform for RawText {
        fn print(&mut self, c: char) {
            self.0.push(c);
        }

        fn execute(&mut self, _: u8) {}

        fn hook(&mut self, _: &vte::Params, _: &[u8], _: bool, _: char) {}

        fn put(&mut self, b: u8) {
            self.0.push(b as char);
        }

        fn unhook(&mut self) {}

        fn osc_dispatch(&mut self, _: &[&[u8]], _: bool) {}

        fn csi_dispatch(&mut self, _: &vte::Params, _: &[u8], _: bool, _: char) {}

        fn esc_dispatch(&mut self, _: &[u8], _: bool, _: u8) {}
    }

    let mut raw = String::new();
    for line in output.lines() {
        let mut performer = RawText::default();
        let mut p = vte::Parser::new();
        for b in line.as_bytes() {
            p.advance(&mut performer, *b);
        }
        raw.push_str(&performer.0);
        raw.push('\n');
    }

    // Remove absolute paths from snapshot tests
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    let manifest_dir: PathBuf = PathBuf::from(manifest_dir);
    let parent = manifest_dir.parent().unwrap();
    let result = raw.replace(&format!("{}/", parent.display()), "");

    // Remove compilation time
    let r = Regex::new("(Finished (debug|release) \\[.*?\\] target\\(s\\) \\[.*?\\] in )(.*?s)")
        .unwrap();
    let result = r.replace(&result, "$1???");

    result.to_string()
}
