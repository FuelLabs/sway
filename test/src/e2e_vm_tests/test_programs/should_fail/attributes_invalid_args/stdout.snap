---
source: test/src/snapshot/mod.rs
assertion_line: 125
---
> forc build --path test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args
exit status: 1
output:
    Building test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args
   Compiling library attributes_invalid_args (test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args)
warning: Attribute is unknown
 --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/unknown_attr.sw:3:3
  |
...
3 | #[unknown(arg_1)]
  |   ------- "unknown" attribute is unknown.
  |
  = help: Unknown attributes are allowed and can be used by third-party tools,
  = help: but the compiler ignores them.
____

warning: Attribute is unknown
 --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/unknown_attr.sw:4:3
  |
...
4 | #[unknown(arg_1, arg_2)]
  |   ------- "unknown" attribute is unknown.
  |
  = help: Unknown attributes are allowed and can be used by third-party tools,
  = help: but the compiler ignores them.
____

warning: Attribute argument is unknown
 --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/allow_attr.sw:9:9
  |
...
9 | #[allow(ded_code)]
  |         -------- "ded_code" is an unknown argument for attribute "allow".
  |         -------- help: Did you mean "dead_code"?
  |         -------- help: Known arguments are "dead_code" and "deprecated".
  |
  = help: Unknown attribute arguments are allowed for some attributes like "allow".
  = help: They can be used by third-party tools, but the compiler ignores them.
____

warning: Attribute argument is unknown
  --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/allow_attr.sw:10:9
   |
...
10 | #[allow(unknown_arg_1, unknown_arg_2)]
   |         ------------- "unknown_arg_1" is an unknown argument for attribute "allow".
   |         ------------- help: Known arguments are "dead_code" and "deprecated".
   |
   = help: Unknown attribute arguments are allowed for some attributes like "allow".
   = help: They can be used by third-party tools, but the compiler ignores them.
____

warning: Attribute argument is unknown
  --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/allow_attr.sw:10:24
   |
...
10 | #[allow(unknown_arg_1, unknown_arg_2)]
   |                        ------------- "unknown_arg_2" is an unknown argument for attribute "allow".
   |                        ------------- help: Known arguments are "dead_code" and "deprecated".
   |
   = help: Unknown attribute arguments are allowed for some attributes like "allow".
   = help: They can be used by third-party tools, but the compiler ignores them.
____

error: Attribute can be applied only once
  --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/storage_attr.sw:10:7
   |
...
 9 |     #[storage(red, writte)]
   |       ------- info: It is already applied here.
10 |     #[storage(unknown_arg)]
   |       ^^^^^^^ "storage" attribute can be applied only once, but is applied two times.
   |
____

error: Attribute argument is invalid
 --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/storage_attr.sw:9:15
  |
...
9 |     #[storage(red, writte)]
  |               ^^^ "red" is an invalid argument for attribute "storage".
  |               --- help: Did you mean "read"?
  |               --- help: Valid arguments are "read" and "write".
  |
____

error: Attribute argument is invalid
 --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/storage_attr.sw:9:20
  |
...
9 |     #[storage(red, writte)]
  |                    ^^^^^^ "writte" is an invalid argument for attribute "storage".
  |                    ------ help: Did you mean "write"?
  |                    ------ help: Valid arguments are "read" and "write".
  |
____

error: Attribute argument is invalid
  --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/storage_attr.sw:10:15
   |
...
10 |     #[storage(unknown_arg)]
   |               ^^^^^^^^^^^ "unknown_arg" is an invalid argument for attribute "storage".
   |               ----------- help: Valid arguments are "read" and "write".
   |
____

error: Attribute can be applied only once
  --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/inline_attr.sw:13:7
   |
...
12 |     #[inline(alwys)]
   |       ------ info: It is already applied here.
13 |     #[inline(unknown_arg)]
   |       ^^^^^^ "inline" attribute can be applied only once, but is applied two times.
   |
____

error: Attribute argument is invalid
  --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/inline_attr.sw:12:14
   |
...
12 |     #[inline(alwys)]
   |              ^^^^^ "alwys" is an invalid argument for attribute "inline".
   |              ----- help: Did you mean "always"?
   |              ----- help: Valid arguments are "always" and "never".
   |
____

error: Attribute argument is invalid
  --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/inline_attr.sw:13:14
   |
...
13 |     #[inline(unknown_arg)]
   |              ^^^^^^^^^^^ "unknown_arg" is an invalid argument for attribute "inline".
   |              ----------- help: Valid arguments are "always" and "never".
   |
____

error: Attribute can be applied only once
 --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/test_attr.sw:7:3
  |
...
6 | #[test(should_rvert)]
  |   ---- info: It is already applied here.
7 | #[test(unknown_arg)]
  |   ^^^^ "test" attribute can be applied only once, but is applied two times.
  |
____

error: Attribute argument is invalid
 --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/test_attr.sw:6:8
  |
...
6 | #[test(should_rvert)]
  |        ^^^^^^^^^^^^ "should_rvert" is an invalid argument for attribute "test".
  |        ------------ help: Did you mean "should_revert"?
  |        ------------ help: The only valid argument is "should_revert".
  |
____

error: Attribute argument is invalid
 --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/test_attr.sw:7:8
  |
...
7 | #[test(unknown_arg)]
  |        ^^^^^^^^^^^ "unknown_arg" is an invalid argument for attribute "test".
  |        ----------- help: The only valid argument is "should_revert".
  |
____

error: Number of attribute arguments is invalid
 --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/payable_attr.sw:7:15
  |
...
7 |     #[payable(invalid)] // Should be no invalid arg error or warning here.
  |               ^^^^^^^ "payable" attribute must not have any arguments, but has one.
  |
____

error: Attribute argument is invalid
 --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/cfg_attr.sw:9:7
  |
...
9 | #[cfg(trget = "fuel")]
  |       ^^^^^ "trget" is an invalid argument for attribute "cfg".
  |       ----- help: Did you mean "target"?
  |       ----- help: Valid arguments are:
  |       ----- help:   - program_type
  |       ----- help:   - target
  |       ----- help:   - experimental_const_generics
  |       ----- help:   - experimental_error_type
  |       ----- help:   - experimental_new_encoding
  |       ----- help:   - experimental_references
  |
____

error: Attribute can be applied only once
 --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/deprecated_attr.sw:7:3
  |
...
6 | #[deprecated(noe = "note")]
  |   ---------- info: It is already applied here.
7 | #[deprecated(unknown_arg = "note")]
  |   ^^^^^^^^^^ "deprecated" attribute can be applied only once, but is applied two times.
  |
____

error: Attribute argument is invalid
 --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/deprecated_attr.sw:6:14
  |
...
6 | #[deprecated(noe = "note")]
  |              ^^^ "noe" is an invalid argument for attribute "deprecated".
  |              --- help: Did you mean "note"?
  |              --- help: The only valid argument is "note".
  |
____

error: Attribute argument is invalid
 --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/deprecated_attr.sw:7:14
  |
...
7 | #[deprecated(unknown_arg = "note")]
  |              ^^^^^^^^^^^ "unknown_arg" is an invalid argument for attribute "deprecated".
  |              ----------- help: The only valid argument is "note".
  |
____

error: Number of attribute arguments is invalid
 --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/fallback_attr.sw:6:12
  |
...
6 | #[fallback(invalid)] // Should be no invalid arg error or warning here.
  |            ^^^^^^^ "fallback" attribute must not have any arguments, but has one.
  |
____

error: Number of attribute arguments is invalid
 --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/error_type_attr.sw:9:14
  |
...
9 | #[error_type(invalid)] // Should be no invalid arg error or warning here.
  |              ^^^^^^^ "error_type" attribute must not have any arguments, but has one.
  |
____

error: Attribute argument is invalid
  --> test/src/e2e_vm_tests/test_programs/should_fail/attributes_invalid_args/src/error_attr.sw:11:13
   |
...
11 |     #[error(msg = "not ok")]
   |             ^^^ "msg" is an invalid argument for attribute "error".
   |             --- help: Did you mean "m"?
   |             --- help: The only valid argument is "m".
   |
____

  Aborting due to 18 errors.
error: Failed to compile attributes_invalid_args
